builder/docker/ecr_login.go

```diff
@@ -38,10 +40,120 @@ type AwsAccessConfig struct {
 	Profile string `mapstructure:"aws_profile" required:"false"`
 }
 
-// Get a login token for Amazon AWS ECR. Returns username and password
+type ECRType string
+
+const (
+	Public  ECRType = "public"
+	Private ECRType = "private"
+	Invalid ECRType = "invalid"
+)
+
+const EcrPublicHost = "public.ecr.aws"
+
+// EcrPublicApiRegion : The Amazon ECR Public registry requires authentication in the us-east-1 Region,
+// so you need to specify --region us-east-1 each time you authenticate
+const EcrPublicApiRegion = "us-east-1"
+
+// GetEcrType : Get ECR type (Public or Private) based on the given URL.
+// If the URL can't be parsed the function returns Invalid.
+func (c *AwsAccessConfig) GetEcrType(ecrUrl string) (ECRType, error) {
+	_, err := url.ParseRequestURI(ecrUrl)
+	if err != nil {
+		return Invalid, err
+	}
+
+	u, err := url.Parse(ecrUrl)
+	if err != nil || u.Scheme == "" || u.Host == "" {
+		return Invalid, err
+	}
+
+	if u.Host == EcrPublicHost {
+		return Public, nil
+	}
+	return Private, nil
+}
+
+// PublicEcrLogin : Get a login token for Amazon AWS ECR Public. Returns username and password
+// or an error.
+func (c *AwsAccessConfig) PublicEcrLogin(ecrUrl string) (string, string, error) {
+	config := aws.NewConfig().WithCredentialsChainVerboseErrors(true)
+	config = config.WithRegion(EcrPublicApiRegion)
+
+	config = config.WithHTTPClient(cleanhttp.DefaultClient())
+	transport := config.HTTPClient.Transport.(*http.Transport)
+	transport.Proxy = http.ProxyFromEnvironment
+
+	// Figure out which possible credential providers are valid; test that we
+	// can get credentials via the selected providers, and set the providers in
+	// the config.
+	creds, err := c.GetCredentials(config)
+	if err != nil {
+		return "", "", fmt.Errorf(err.Error())
```

Comment:
Agreeing with @lbajolet-hashicorp here. You can return the error up to the calling function. The caller already adds a bit of context indicating that it failed to fetch the ECR credentials. 
```suggestion
		return "", "", err
```